### PR TITLE
3.3 RFC: Transaction Isolation Level Abstraction

### PR DESCRIPTION
**Pull request by half2me**
My idea for RFC #3136 

Still needs lots of work, but I need some help, and wanna make sure I'm heading in the right direction.
@lorenzo ?

### CODE REVIEW

File: src/Database/Connection.php [hunk 1]
```diff
@@ -769,4 +770,9 @@ public function __debugInfo()
             'logger' => $this->_logger
         ];
     }
+
+    private function isolationLevel($isolationLevel = null)
```
Comment (half2me) [inline src/Database/Connection.php]: oops, that was meant to be public :)

File: src/Database/Connection.php [hunk 2]
```diff
@@ -548,11 +549,12 @@ public function supportsDynamicConstraints()
...
+     * }, Connection::SERIALIZABLE);
      * ```
      */
-    public function transactional(callable $callback)
+    public function transactional(callable $callback, $isolationLevel = null)
     {
+        $this->isolationLevel($isolationLevel);
         $this->begin();
 
         try {
```
Comment (lorenzo) [inline src/Database/Connection.php]: Missing a reset for the previous level

File: src/Database/Driver.php [hunk 3]
```diff
@@ -361,4 +362,14 @@ public function __debugInfo()
...
+
+    /**
+     * Returns the session isolationLevel
+     * Sets it, if a value is given
+     *
+     * @param string $isolationLevel Transaction Isolation Level
+     * @return string
+     * @throws TransactionIsolationLevelNotSupportedException
+     */
+    abstract public function transactionIsolationLevel($isolationLevel = null);
```
Comment (lorenzo) [inline src/Database/Driver.php]: cannot have it as abstract in a minor version, since it would break other people's code

**Comment by lorenzo** [comment]
I looks like it is going in the right direction

File: src/Database/Driver/PDODriverTrait.php [hunk 4]
```diff
@@ -153,6 +154,23 @@ public function rollbackTransaction()
...
+     * Returns the session isolationLevel
+     * Sets it, if a value is given
+     *
+     * @param string $isolationLevel Transaction Isolation Level
+     * @return string
+     * @throws TransactionIsolationLevelNotSupportedException
+     */
+    public function transactionIsolationLevel($isolationLevel = null)
+    {
+        if (!is_null($isolationLevel) && strlen($isolationLevel) > 0) {
```
Comment (dereuromark) [inline src/Database/Driver/PDODriverTrait.php]: `!== null`

File: src/Database/Driver.php (hunk 3 above)
Comment (half2me) [inline src/Database/Driver.php]: True

**Comment by markstory** [comment]
ping.

**Comment by half2me** [comment]
@markstory I'll have time to work on it next week. Unfortunately had too much work accumulated to do any work on this recently.

**Comment by half2me** [comment]
any help would be appreciated on this...

**Comment by markstory** [comment]
Closing again as it has been a long time with no activity, and the changes can no longer be merged. If the branch can be made mergeable please reopen this pull request.

**Pull request closed without merging**